crates/sui-adapter/src/programmable_transactions/context.rs

```diff
@@ -0,0 +1,224 @@
+}
+impl<'v, 's, 'a, 'b, E, S> ExecutionContext<'v, 's, 'a, 'b, E, S>
+where
+    E: fmt::Debug,
+    S: ResourceResolver<Error = E>
+        + ModuleResolver<Error = E>
+        + Storage
+        + ParentSync
+        + ChildObjectResolver,
+{
+    pub fn new(
+        vm: &'v MoveVM,
+        state_view: &'s S,
+        ctx: &'a mut TxContext,
+        gas_status: &'a mut GasStatus<'b>,
+        gas_coin: ObjectID,
+        inputs: Vec<CallArg>,
+    ) -> Result<Self, ExecutionError> {
+        let mut object_owner_map = BTreeMap::new();
+        let inputs = inputs
+            .into_iter()
+            .map(|call_arg| {
+                Ok(Some(load_call_arg(
+                    state_view,
+                    &mut object_owner_map,
+                    call_arg,
+                )?))
+            })
+            .collect::<Result<_, ExecutionError>>()?;
+        let gas = Some(Value::Object(load_object(
+            state_view,
+            &mut object_owner_map,
+            gas_coin,
+        )?));
+        Ok(Self {
+            vm,
+            state_view,
+            ctx,
+            gas_status,
+            session: None,
+            object_owner_map,
+            gas,
+            inputs,
+            results: vec![],
+            additional_transfers: vec![],
+        })
+    }
+
+    pub fn session(&mut self) -> &Session<'s, 'v, S> {
+        self.session.get_or_insert_with(|| {
+            new_session(
+                self.vm,
+                self.state_view,
+                self.object_owner_map.clone(),
+                self.gas_status.is_metered(),
+            )
+        })
+    }
+
+    pub fn fresh_id(&mut self) -> Result<ObjectID, ExecutionError> {
+        if true {
+            todo!("update native context set")
+        }
+        Ok(self.ctx.fresh_id())
+    }
+
+    pub fn delete_id(&mut self, _object_id: ObjectID) -> Result<(), ExecutionError> {
+        if true {
+            todo!("update native context set")
+        }
+        Ok(())
+    }
+
+    pub fn take_args<V: TryFromValue>(
+        &mut self,
+        args: Vec<Argument>,
+    ) -> Result<Vec<V>, ExecutionError> {
+        args.into_iter()
+            .map(|arg| self.take_arg::<V>(arg))
+            .collect()
+    }
+
+    pub fn take_arg<V: TryFromValue>(&mut self, arg: Argument) -> Result<V, ExecutionError> {
+        let val_opt = self.borrow_mut(arg)?;
+        if val_opt.is_none() {
+            panic!("taken value")
+        }
+        V::try_from_value(val_opt.take().unwrap())
+    }
+
+    pub fn restore_arg(&mut self, arg: Argument, value: Value) -> Result<(), ExecutionError> {
+        let val_opt = self.borrow_mut(arg)?;
+        assert_invariant!(
+            val_opt.is_none(),
+            "Should never restore a non-taken value. \
+            The take+restore is an implementation detail of mutable references"
+        );
+        *val_opt = Some(value);
+        Ok(())
```

Comment:
Could use `replace` here?  It's perhaps slightly cleaner, but then I'm not sure what the implications are in the error case, because the previous occupant of the argument will still have been overwritten.  I'm only suggesting it because I noticed that in `load_object`, you do something similar with `insert` for the object owner mapping

```suggestion
        let old_val = self.borrow_mut(arg)?.replace(value);
        assert_invariant!(
            old_val.is_none(),
            "Should never restore a non-taken value. \
            The take+restore is an implementation detail of mutable references"
        );
        Ok(())
```